### PR TITLE
Feature/device pairing

### PR DESCRIPTION
**Pull request by S-Traut**
Adding a new command to be able to pair bragi devices.
This as been tested with a **CORSAIR IRONCLAW WIRELESS USB MOUSE**

Usage:
```
echo pair ckbX,ckbX > /dev/input/ckbX/cmd
Where left side is the devices you want to pair (mouse/keyboard/...) and right side is the dongle device
```

Workflow used to pair the devices:
```
Generates a new PAIRING_ID (8B)
Generates a new ENCRYPTION_KEY (16B)
Set on the dongle device the pairing id and the encryption key with 2 bragi_write_to_handle executions on 0x5 and 0x6 
(the encryption key must be set exactly after the pairing id otherwise it will no work.)
Set each mouse/keyboard devices precised in parameters the same Pairing ID and Encryption key in the exact same way.
```

Note that this is only a command that can be used through CLI command. There is no GUI integration yet.

The encryption key generation as been done through the /dev/urandom in ``src/daemon/utils.c``

### CODE REVIEW

File: src/daemon/usb.h
Comment (S-Traut) [inline src/daemon/usb.h:1]: @tatokis What should we do about this?

**Comment by S-Traut** [comment]
link to #920

File: src/daemon/CMakeLists.txt [hunk 1]
```diff
@@ -119,6 +119,8 @@ if (MACOS OR LINUX)
               usb_bragi.h
               bragi_common.h
               bragi_notification.h
+              utils.h
+              utils.c
```
Comment (tatokis) [inline src/daemon/CMakeLists.txt:123]: Please move `utils.c` up along with the rest of the .c files.

File: src/daemon/command.c [hunk 2]
```diff
@@ -136,7 +137,7 @@ int readcmd(usbdevice* kb, char* line){
             continue;
         }
         // Reject anything not related to fwupdate if device has a bricked FW
-        if(NEEDS_FW_UPDATE(kb) && command != FWUPDATE && command != NOTIFYON && command != NOTIFYOFF && command != RESET)
+        if(NEEDS_FW_UPDATE(kb) && command != FWUPDATE && command != NOTIFYON && command != NOTIFYOFF && command != RESET && command != PAIR)
```
Comment (tatokis) [inline src/daemon/command.c:140]: As per our discussion, this isn't needed, as we can't pair devices in bootloader mode anyway.

File: src/daemon/device_bragi.c [hunk 3]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
     // Since we use usbrecv for everything, the devices tell us when they are ready to handle another packet.
     return;
 }
+
+static void bragi_generate_encryption_key(uchar* encryption_key) {
```
Comment (tatokis) [inline src/daemon/device_bragi.c:287]: 4 space indentation please.

File: src/daemon/device_bragi.c [hunk 4]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
...
+  ckb_generate_random(16, encryption_key);
+  for(int i = 0; i < 16; i++) {
```
Comment (tatokis) [inline src/daemon/device_bragi.c:289]: This should be removed. We did identify that iCUE sends bytes > 127.

File: src/daemon/device_bragi.c [hunk 5]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
...
+
+  uchar pairing_id[8];
+  ckb_generate_random(8, pairing_id);
+  ckb_info("Generated new pairing id");
+
+  // Set the dongle device pairing_id and encryption_key
+  bragi_set_device_keys(kb, pairing_id, encryption_key);    
+
+  char* devices = strdup(to);
+  char *token_ptr;
```
Comment (tatokis) [inline src/daemon/device_bragi.c:328]: `char* token_ptr`

File: src/daemon/device_bragi.c [hunk 6]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
...
+  char* device_name = strtok_r(devices, ",", &token_ptr);
+  while(device_name != NULL) {
+    
+    int ckb_id = 0;
+    if(!sscanf(device_name, "ckb%d", &ckb_id)) {
+      ckb_err("Failed to parse device name: %s", device_name);
+      return;
+    }
+
+    usbdevice* device = keyboard + ckb_id;
```
Comment (tatokis) [inline src/daemon/device_bragi.c:338]: Please lock the dmutex before accessing any part of `device`.

File: src/daemon/device_bragi.c [hunk 7]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
...
+    
+    int ckb_id = 0;
+    if(!sscanf(device_name, "ckb%d", &ckb_id)) {
+      ckb_err("Failed to parse device name: %s", device_name);
+      return;
+    }
+
+    usbdevice* device = keyboard + ckb_id; 
+    if(device->status != DEV_STATUS_CONNECTED) {
+      ckb_err("Device not found or connected");
```
Comment (tatokis) [inline src/daemon/device_bragi.c:340]: Maybe specify which device was not found (probably by ckb%d).

File: src/daemon/device_bragi.c [hunk 8]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
...
+  bragi_close_handle(kb, BRAGI_GENERIC_HANDLE);
+}
+
+void cmd_pair_bragi(usbdevice* kb, usbmode* mode, int dummy1, int dummy2, const char* to) {
+  (void)mode;
+  (void)dummy1;
+  (void)dummy2;
+
+  uchar encryption_key[16];
+  bragi_generate_encryption_key(encryption_key);
```
Comment (tatokis) [inline src/daemon/device_bragi.c:317]: Should probably check to see if the generation succeeded.

File: src/daemon/device_bragi.c [hunk 9]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
...
+  ckb_info("Generated new encryption key");
+
+  uchar pairing_id[8];
+  ckb_generate_random(8, pairing_id);
```
Comment (tatokis) [inline src/daemon/device_bragi.c:321]: Likewise

File: src/daemon/utils.c [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#include "utils.h"
+#include "includes.h"
+
+void ckb_generate_random(unsigned int size, unsigned char* data) {
```
Comment (tatokis) [inline src/daemon/utils.c:4]: Could directly return ssize_t to indicate success/failure (and also perhaps alert in case of short reads).

File: src/daemon/utils.c [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  int rand = open("/dev/urandom", O_RDONLY);
```
Comment (tatokis) [inline src/daemon/utils.c:5]: I think I'd prefer /dev/random

File: src/daemon/utils.c [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  if (rand < 0)
+    ckb_err("Error loading /dev/urandom");
+    
+  ssize_t result = read(rand, data, sizeof(unsigned char) * size);
+
+  if (result < 0)
+    ckb_err("Error generating encryption key");
```
Comment (tatokis) [inline src/daemon/utils.c:12]: "Generating random data", as this is used for pairing ID too

File: src/daemon/utils.c (hunk 10 above)
Comment (tatokis) [inline src/daemon/utils.c:4]: Also, I think size_t is more appropriate for the size parameter. Then it can be called as such: ckb_generate_random(sizeof(whatever_buffer), whatever_buffer);

File: src/daemon/device_bragi.c [hunk 13]
```diff
@@ -281,3 +283,69 @@ void bragi_delay(usbdevice* kb, delay_type_t type){
...
+
+  uchar pkt[BRAGI_JUMBO_SIZE] = {0};
+
+  // Write the PAIRING_ID to the device first
+  bragi_open_handle(kb, BRAGI_GENERIC_HANDLE, BRAGI_RES_PAIRINGID);
+  memcpy(pkt + 7, pairing_id, sizeof(uchar) * 8);
+  bragi_write_to_handle(kb, pkt, BRAGI_GENERIC_HANDLE, sizeof(pkt), sizeof(uchar) * 8);
+  bragi_close_handle(kb, BRAGI_GENERIC_HANDLE);
+
+  // The ENCRYPTION_KEY must be writting directly after the PAIRING_ID
```
Comment (tatokis) [inline src/daemon/device_bragi.c:304]: "be written"

File: src/daemon/device_bragi.c [hunk 14]
```diff
@@ -233,6 +234,7 @@ int start_dongle_bragi(usbdevice* kb, int makeactive){
     // Force back to SW mode
     // FIXME: Does this make a difference?
     bragi_set_property(kb, BRAGI_MODE, BRAGI_MODE_SOFTWARE);
+    kb->active = 1;
```
Comment (tatokis) [inline src/daemon/device_bragi.c:237]: This should probably be a separate commit (before the one that introduces the pair commands)

**Review by tatokis — CHANGES_REQUESTED** [review]
Please rebase and squash commits as needed. Ideally in my opinion:
1. Mark bragi dongles correctly as active
2. Implement pair command

Although this is just a guideline, I'd like to keep any chances that have even the slightest chance to regress something as separate as possible (so that it can be easily git bisected down the line).

Once the feedback is addressed, I'll test it on my own hardware.

File: src/daemon/usb.h
Comment (tatokis) [reply src/daemon/usb.h:1]: Please remove this change from this PR. If you want, you can try adding the generic dongle as a separate device (without marking it as a mouse) and see if it still works. If it does, you can open a new PR. Basically look at where the existing dongle define is used, and add a new one in there too.

File: src/daemon/device_bragi.c (hunk 8 above)
Comment (S-Traut) [reply src/daemon/device_bragi.c:316]: Just adding a ckb_info call, or do you have some specific checks in mind to assure that this succeeded?

File: src/daemon/device_bragi.c (hunk 6 above)
Comment (S-Traut) [reply src/daemon/device_bragi.c:337]: I'm not very familiar with "dmutex", can you point me out some documentation / tips about it?

File: src/daemon/utils.c [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    if (rand < 0) {
+        ckb_err("Error loading /dev/random");
+        return 0;
+    }
+    
+    ssize_t result = read(rand, data, sizeof(unsigned char) * size);
+
+    if (result < 0) {
+        ckb_err("Error generating random data");
+        return result;
```
Comment (tatokis) [inline src/daemon/utils.c:15]: close() the fd here before returning

**Review by tatokis — CHANGES_REQUESTED** [review]
